src/mlpack/core/data/save_impl.hpp

```diff
@@ -38,220 +39,177 @@ bool Save(const std::string& filename,
+  return Save(filename, matrix, copyOpts);
 }
 
-// Save a Sparse Matrix
-template<typename eT>
+/*
+ * Add this SFINAE in here because the compiler is so stupid that it is not
+ * able to distinguish between these two:
+ *
+ *  data::Save(filename, "model", *output);
+ *
+ *  and 
+ *
+ *  data::Save(filename, matrix, opts);
+ *  
+ * The second SFINAE is added because the compiler is bot able to see the
+ * difference between:
+ *
+ *  data::Save(filename, Row/Col, fatal);
+ *
+ *  and 
+ *
+ *  data::Save(filename, Row/Col, Opts);
+ *
+ * This SFINAE is temporary and must be removed after the integration of stage 3 or
+ * when the compiler becomes more intelligent.
+ */
+template<typename MatType, typename DataOptionsType>
 bool Save(const std::string& filename,
-          const arma::SpMat<eT>& matrix,
-          const bool fatal,
-          bool transpose)
+          const MatType& matrix,
+          DataOptionsType& opts,
+          std::enable_if_t<IsArma<MatType>::value || IsSparseMat<MatType>::value>*,
+          std::enable_if_t<!std::is_same_v<DataOptionsType, bool>>*)
 {
   Timer::Start("saving_data");
 
-  // First we will try to discriminate by file extension.
-  std::string extension = Extension(filename);
-  if (extension == "")
+  bool success = DetectFileType<MatType>(filename, opts, false);
+  if (!success)
   {
     Timer::Stop("saving_data");
-    if (fatal)
-      Log::Fatal << "No extension given with filename '" << filename << "'; "
-          << "type unknown.  Save failed." << std::endl;
-    else
-      Log::Warn << "No extension given with filename '" << filename << "'; "
-          << "type unknown.  Save failed." << std::endl;
-
     return false;
   }
 
-  // Catch errors opening the file.
   std::fstream stream;
-#ifdef  _WIN32 // Always open in binary mode on Windows.
-  stream.open(filename.c_str(), std::fstream::out | std::fstream::binary);
-#else
-  stream.open(filename.c_str(), std::fstream::out);
-#endif
-  if (!stream.is_open())
+  success = OpenFile(filename, opts, false, stream);
+  if (!success)
   {
     Timer::Stop("saving_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "' for writing. "
-          << "Save failed." << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "' for writing; save "
-          << "failed." << std::endl;
-
     return false;
   }
 
-  bool unknownType = false;
-  FileType saveType;
-  std::string stringType;
-
-  if (extension == "txt" || extension == "tsv")
-  {
-    saveType = FileType::CoordASCII;
-    stringType = "raw ASCII formatted data";
-  }
-  else if (extension == "bin")
+  // Try to save the file.
+  Log::Info << "Saving " << opts.FileTypeToString() << " to '" << filename
+      << "'." << std::endl;
+  if constexpr (IsArma<MatType>::value || IsSparseMat<MatType>::value)
   {
-    saveType = FileType::ArmaBinary;
-    stringType = "Armadillo binary formatted data";
+    TextOptions txtOpts(std::move(opts));
```

Comment:
To make `opts` const, we would probably just have to remove the `std::move` here... but we could also remove the line `opts = std::move(txtOpts)` down below.